### PR TITLE
Better anonymous record error reporting

### PR DESCRIPTION
**Pull request by edgarfgp**
Fixes https://github.com/dotnet/fsharp/issues/14814

Update: A new quick fix can be added based on the new error number

### CODE REVIEW

File: src/Compiler/Checking/NameResolution.fs [hunk 1]
```diff
@@ -3685,8 +3685,9 @@ let ResolveFieldPrim sink (ncenv: NameResolver) nenv ad ty (mp, id: Ident) allFi
     | [] ->
         let lookup() =
             let frefs =
-                try Map.find id.idText nenv.eFieldLabels
-                with :? KeyNotFoundException ->
+                match Map.tryFind id.idText nenv.eFieldLabels with
+                | Some fields -> fields
+                | None ->
```
[the remark this answers is not in the document]
Comment (edgarfgp) [reply src/Compiler/Checking/NameResolution.fs:3690]: I do not think it changes anything. I just wanted to void using `try with`, but this is an unrelated change I guess I can revert this to make more easy to review.

File: src/Compiler/Checking/NameResolution.fs [hunk 2]
```diff
@@ -3749,16 +3749,26 @@ let ResolveFieldPrim sink (ncenv: NameResolver) nenv ad ty (mp, id: Ident) allFi
             errorR(Error(FSComp.SR.nrInvalidFieldLabel(), (List.head rest).idRange))
 
         [(resInfo, item)]
+        
+let ResolveAnonRecField (g: TcGlobals) ty (fldId: Ident) =
+    if (TryFindAnonRecdFieldOfType g ty fldId.idText).IsSome then
+        error(Error(FSComp.SR.chkAnonymousRecordFields(fldId.idText, fldId.idText), fldId.idRange))
```
[the remark this answers is not in the document]
Comment (edgarfgp) [reply src/Compiler/Checking/NameResolution.fs:3755]: Will do. this is still WIP as Im getting familiar with this part the compiler :)

File: src/Compiler/FSComp.txt [hunk 3]
```diff
@@ -1700,5 +1700,6 @@ featureInformationalObjInferenceDiagnostic,"Diagnostic 3559 (warn when obj infer
...
 3567,parsMissingMemberBody,"Expecting member body"
 3568,parsMissingKeyword,"Missing keyword '%s'"
-3569,chkNotTailRecursive,"The member or function '%s' has the 'TailCallAttribute' attribute, but is not being used in a tail recursive way."
-3577,tcOverrideUsesMultipleArgumentsInsteadOfTuple,"This override takes a tuple instead of multiple arguments. Try to add an additional layer of parentheses at the method definition (e.g. 'member _.Foo((x, y))'), or remove parentheses at the abstract method declaration (e.g. 'abstract member Foo: 'a * 'b -> 'c')."
\ No newline at end of file
+3577,tcOverrideUsesMultipleArgumentsInsteadOfTuple,"This override takes a tuple instead of multiple arguments. Try to add an additional layer of parentheses at the method definition (e.g. 'member _.Foo((x, y))'), or remove parentheses at the abstract method declaration (e.g. 'abstract member Foo: 'a * 'b -> 'c')."
+3578,chkAnonymousRecordFields,"Label '%s' is part of anonymous record. Use {{| expr with %s = ... |}} instead."
```
[the remark this answers is not in the document]
Comment (edgarfgp) [reply src/Compiler/FSComp.txt]: Updated to `chkCopyUpdateSyntaxInAnonRecords` to make it more clear.

File: src/Compiler/Checking/NameResolution.fs (hunk 2 above)
[the remark this answers is not in the document]
Comment (edgarfgp) [reply src/Compiler/Checking/NameResolution.fs]: Updated. Let me know if there is anything extra that I should cover :)

File: src/Compiler/FSComp.txt [hunk 4]
```diff
@@ -1708,4 +1708,5 @@ featureAccessorFunctionShorthand,"underscore dot shorthand for accessor only fun
...
 3570,tcStaticBindingInExtrinsicAugmentation,"Static bindings cannot be added to extrinsic augmentations. Consider using a 'static member' instead."
 3571,pickleFsharpCoreBackwardsCompatible,"Newly added pickle state cannot be used in FSharp.Core, since it must be working in older compilers+tooling as well. The time window is at least 3 years after feature introduction. Violation: %s . Context: \n %s "
-3577,tcOverrideUsesMultipleArgumentsInsteadOfTuple,"This override takes a tuple instead of multiple arguments. Try to add an additional layer of parentheses at the method definition (e.g. 'member _.Foo((x, y))'), or remove parentheses at the abstract method declaration (e.g. 'abstract member Foo: 'a * 'b -> 'c')."
\ No newline at end of file
+3577,tcOverrideUsesMultipleArgumentsInsteadOfTuple,"This override takes a tuple instead of multiple arguments. Try to add an additional layer of parentheses at the method definition (e.g. 'member _.Foo((x, y))'), or remove parentheses at the abstract method declaration (e.g. 'abstract member Foo: 'a * 'b -> 'c')."
+3578,chkCopyUpdateSyntaxInAnonRecords,"Label '%s' is part of anonymous record. Use {{| %s with %s = ... |}} instead."
```
Comment (T-Gro) [inline src/Compiler/FSComp.txt:1712]:
```suggestion
3578,chkCopyUpdateSyntaxInAnonRecords,"Value '%s' is an anonymous record. Use syntax {{| %s with %s = ... |}} to update it."
```

+ switch the arguments.
Imo the name of the anon record value should  be the subject in this sentence.

Comment (vzarytovskii) [reply src/Compiler/FSComp.txt:1712]: @T-Gro, probably "use the following syntax ... to update it"? My english is not very good, maybe @KathleenDollard can help us here?

Comment (edgarfgp) [reply src/Compiler/FSComp.txt:1712]: Happy to adjust the error message. I was hoping for a better error suggested :).

**Comment by psfinaki** [comment]
> Update: A new quick fix can be added based on the new error number

Actually we already have the fix "Convert To Anonymous Record" which happens to handle this correctly currently. After this, it will stop working for that case. The tests we have now are handling a different FS0039 case (which still holds) hence nothing is failing in this PR.

If you have a chance, you can try adding the new diag [here](https://github.com/dotnet/fsharp/blob/3d15dd91233d68a118d0e0746463e883ff688089/vsintegration/src/FSharp.Editor/CodeFixes/ConvertToAnonymousRecord.fs#L19C69-L19C69) and the respective test [here](https://github.com/dotnet/fsharp/blob/main/vsintegration/tests/FSharp.Editor.Tests/CodeFixes/ConvertToAnonymousRecordTests.fs) and see what happens.

**Comment by edgarfgp** [comment]
> > Update: A new quick fix can be added based on the new error number
> 
> Actually we already have the fix "Convert To Anonymous Record" which happens to handle this correctly currently. After this, it will stop working for that case. The tests we have now are handling a different FS0039 case (which still holds) hence nothing is failing in this PR.
> 
> If you have a chance, you can try adding the new diag [here](https://github.com/dotnet/fsharp/blob/3d15dd91233d68a118d0e0746463e883ff688089/vsintegration/src/FSharp.Editor/CodeFixes/ConvertToAnonymousRecord.fs#L19C69-L19C69) and the respective test [here](https://github.com/dotnet/fsharp/blob/main/vsintegration/tests/FSharp.Editor.Tests/CodeFixes/ConvertToAnonymousRecordTests.fs) and see what happens.

I see. I think `FS0039` is used in many cases where a `Label` is not defined, hence assuming that the fix will be a `ConvertToAnonymousRecord`. This new error code will check really make sure we are in the context of an `AnonymousRecord`.

 I think this new error code is more accurate to when we should offer a `ConvertToAnonymousRecord` quick fix.

File: tests/service/Symbols.fs [hunk 5]
```diff
@@ -751,3 +752,17 @@ type Foo() =
...
             (5, 14, "    member _.X", "X")
+            
+module AnonymousRecord =
+    [<Test>]
+    let ``Anonymous record copy-and-update symbols`` () =
+        let _, checkResults = getParseAndCheckResults """
+let f (r: {| A: int; C: int |}) =
+    { r with A = 1; B = 2; C = 3 }
+"""
+        assertHasSymbolUsages [ "A"; "C" ] checkResults
```
Comment (auduchinok) [inline tests/service/Symbols.fs:763]: This doesn't actually check the usages of `A`, and `C`. Since they are declared in the type annotation, they will make this check pass regardless of whether the actual usages are reported below.

File: src/Compiler/Checking/NameResolution.fs [hunk 6]
```diff
@@ -3685,8 +3685,8 @@ let ResolveFieldPrim sink (ncenv: NameResolver) nenv ad ty (mp, id: Ident) allFi
     | [] ->
         let lookup() =
             let frefs =
-                try Map.find id.idText nenv.eFieldLabels
-                with :? KeyNotFoundException ->
+                 try Map.find id.idText nenv.eFieldLabels
+                 with :? KeyNotFoundException ->
```
Comment (auduchinok) [inline src/Compiler/Checking/NameResolution.fs:3689]: Could you revert these changes, please?

File: src/Compiler/Checking/CheckExpressions.fs [hunk 7]
```diff
@@ -7388,7 +7400,11 @@ and TcRecdExpr cenv overallTy env tpenv (inherits, withExprOpt, synRecdFields, m
         match flds with
         | [] -> []
         | _ ->
-            match BuildFieldMap cenv env hasOrigExpr overallTy flds mWholeExpr with
+            let tyIdent =
+                match withExprOpt with
+                | Some (SynExpr.Ident ident, _) -> Some ident
```
Comment (auduchinok) [inline src/Compiler/Checking/CheckExpressions.fs:7405]: Why does it check a specific shape of the tree and only works for a simple identifier?

It's possible to get an anonymous property from a property (`x.Prop`), a function call (`f ()`), an indexer (`x.[0]`, or `x[0]`), and so on. There are many more different shapes of a tree that we may get here, and we should not assume some specific one, nor should we hardcode a few popular ones, while disabling recovery and analysis for all other cases.

File: src/Compiler/FSComp.txt (hunk 4 above)
Comment (auduchinok) [reply src/Compiler/FSComp.txt:1712]: Do we actually use term `Label` for record fields anywhere in the docs?

**Review by auduchinok — COMMENTED** [review]
Assuming a specific tree shape during a general type checking phase doesn't seem right thing to do in a compiler. Please make it more generic.

File: src/Compiler/Checking/CheckExpressions.fs (hunk 7 above)
Comment (edgarfgp) [reply src/Compiler/Checking/CheckExpressions.fs:7405]: You are right. I was just interested in the `idText`. I will have a look a better way to do this

**Comment by edgarfgp** [comment]
@psfinaki Updated the Quick Fix and added tests and all seems to be working locally

File: src/Compiler/FSComp.txt (hunk 4 above)
Comment (edgarfgp) [reply src/Compiler/FSComp.txt:1712]: @auduchinok I will wait to see what is the final suggested message to update it :)

File: tests/service/Symbols.fs (hunk 5 above)
Comment (edgarfgp) [reply tests/service/Symbols.fs:763]: Ok. let me check if there is any example of how to do that. I'm open to suggestions

**Comment by psfinaki** [comment]
@edgarfgp this is awesome - thanks.

Yeah I am not a big anon records user so cannot really say if a dev would really expect this fix for the case it's working for currently. And indeed triggering the code fix evaluation for every 0039 is kind of costly where the new diag is a more targeted and appropriate application.

That said, I don't mind keeping it for both diags for now until somebody complains about it's speed or uselessness. After all, somebody created it in its current shape so was motivated :D

**Review by psfinaki — APPROVED** [review]

File: src/Compiler/Checking/CheckExpressions.fs (hunk 7 above)
Comment (edgarfgp) [reply src/Compiler/Checking/CheckExpressions.fs]: Updated as discussed

File: tests/service/Symbols.fs (hunk 5 above)
Comment (edgarfgp) [reply tests/service/Symbols.fs]: Updated

File: tests/service/Symbols.fs [hunk 8]
```diff
@@ -751,3 +752,32 @@ type Foo() =
...
+let f (x: {| A: int |}) =
+    {| x with A = 1 |}
+"""
+
+        let getSymbolUses =
+            checkResults.GetAllUsesOfAllSymbolsInFile()
+            |> Seq.filter (fun su -> su.IsFromType)
+            |> Array.ofSeq
+
+        Assert.AreEqual(2, getSymbolUses.Length)
```
Comment (auduchinok) [inline tests/service/Symbols.fs:769]: Are you sure these symbols are not `A` and `int`?

Comment (edgarfgp) [reply tests/service/Symbols.fs:769]: Yeap. you are right as usual. I will try again :)

**Review by auduchinok — CHANGES_REQUESTED** [review]

File: tests/service/Symbols.fs (hunk 5 above)
Comment (auduchinok) [reply tests/service/Symbols.fs]: Previously it tested normal record syntax, and now these tests check anon records syntax instead. Please don't remove tests for the normal syntax.

File: tests/service/Symbols.fs [hunk 9]
```diff
@@ -751,3 +752,52 @@ type Foo() =
...
+    [<Test>]
+    let ``Anonymous record copy-and-update symbols usage`` () =
+        let _, checkResults = getParseAndCheckResults """
+module X
+let f (x: {| A: int |}) =
+    {| x with A = 1 |}
+"""
+
+        let fstSymbol =
+            checkResults.GetSymbolUsesAtLocation(3, 14, "    {| x with A = 1 |}", ["A"])
```
[the remark this answers is not in the document]
Comment (edgarfgp) [reply tests/service/Symbols.fs]: Not a particular reason. Updated the test in my last commit to use the same approach.

**Review by auduchinok — APPROVED** [review]

File: tests/service/Symbols.fs (hunk 5 above)
Comment (edgarfgp) [reply tests/service/Symbols.fs]: Done

Comment (auduchinok) [reply tests/service/Symbols.fs]: @edgarfgp Thanks!

**Comment by edgarfgp** [comment]
Addresses all the comments. Let me know if there is anything else that I need to cover. ?

File: src/Compiler/FSComp.txt (hunk 4 above)
Comment (0101) [reply src/Compiler/FSComp.txt]: This is still a bit confusing for me, because the problem isn't with `'ff'` but rather with `expr`. I think Tomas' proposal makes more sense. Is it possible to get the name of `expr` here? 

Also if the user is already using correct `with` syntax, we could be even more focused on the actual problem, e.g. `'expr' is an anonymous record, use {|...|} instead of {...}`.

File: tests/FSharp.Compiler.ComponentTests/Conformance/Types/RecordTypes/AnonymousRecords.fs [hunk 10]
```diff
@@ -19,7 +19,18 @@ module AnonRecd =
...
+    
+    [<Fact>]
+    let ``Anonymous Records incomplete`` () =
+        Fsx """
+let x () : {| A: int; B: string  |} =  {| A = 123 |}
+"""
+        |> compile
+        |> shouldFail
+        |> withDiagnostics [
+            (Error 1, Line 2, Col 40, Line 2, Col 53, "This anonymous record does not have enough fields. Add the missing fields [B].")
```
Comment (0101) [inline tests/FSharp.Compiler.ComponentTests/Conformance/Types/RecordTypes/AnonymousRecords.fs:31]: Looks like this error message isn't part of this PR, but I think it could also be improved. `enough` seems wrong here, because just having enough fields wouldn't necessarily help if they aren't the correct ones. It should rather be `... doesn't have all <expected|required|defined|specified> fields ...`. Or even just `This anonymous record missing field B` / `This anonymous record missing fields [B, C]`

Not critical though 😄

File: src/Compiler/FSComp.txt (hunk 4 above)
Comment (edgarfgp) [reply src/Compiler/FSComp.txt]: > 

In my initial implementation I had the `Ident` name from `expr` but as Eugene pointed out [here](https://github.com/dotnet/fsharp/pull/15598#discussion_r1272201892). I was only checking for one specific shape of the `SynExpr` and that will eliminate the error reporting for the rest of the cases where an anonymous record value can be obtained like from a property (x.Prop), a function call (f ()), an indexer (x.[0], or x[0]), etc. 

I think are more general approach needs to be used here so that is why I'm using `expr`

File: tests/FSharp.Compiler.ComponentTests/Conformance/Types/RecordTypes/AnonymousRecords.fs (hunk 10 above)
Comment (edgarfgp) [reply tests/FSharp.Compiler.ComponentTests/Conformance/Types/RecordTypes/AnonymousRecords.fs:31]: I can do in a separate PR 🙂

File: src/Compiler/FSComp.txt (hunk 4 above)
Comment (0101) [reply src/Compiler/FSComp.txt]: I suppose we can say `This expression is an anonymous record, use {|...|} instead of {...}.` if the whole `{ ... }` is underlined. I would prefer that over mentioning a particular field from the record. What do others think?

Comment (edgarfgp) [reply src/Compiler/FSComp.txt]: This PR uses a more accurate error range based on the annon record field that is being used.

Comment (0101) [reply src/Compiler/FSComp.txt]: That doesn't seem very accurate to me, seeing that the fields have nothing to do with the error.

File: tests/FSharp.Compiler.ComponentTests/Conformance/Types/RecordTypes/AnonymousRecords.fs [hunk 11]
```diff
@@ -65,7 +76,55 @@ type ErrorResponse =
...
+    let ``This expression was expected to have an anonymous with various properties Record but has a record`` () =
+        Fsx """
+let f (r: {| A: int; C: int |}) = { r with A = 1; B = 2; C = 3 }
+        """
+        |> compile
+        |> shouldFail
+        |> withDiagnostics [
+            (Error 3578, Line 2, Col 44, Line 2, Col 45, "'A' is part of an anonymous record. Use {| expr with A = ... |} instead.")
+            (Error 39, Line 2, Col 51, Line 2, Col 52, "The record label 'B' is not defined.")
+            (Error 3578, Line 2, Col 58, Line 2, Col 59, "'C' is part of an anonymous record. Use {| expr with C = ... |} instead.")
```
Comment (0101) [inline tests/FSharp.Compiler.ComponentTests/Conformance/Types/RecordTypes/AnonymousRecords.fs:125]: Would it make sense to do something like this? Underline the braces (for which we probably have to output 2 diagnostics) and not talk about fields that are correct?

```suggestion
            (Error 3578, Line 2, Col 35, Line 2, Col 35, "This expression is an anonymous record, use {|...|} instead of {...}.")
            (Error 39, Line 2, Col 51, Line 2, Col 52, "The record label 'B' is not defined.")
            (Error 3578, Line 2, Col 64, Line 2, Col 65, "This expression is an anonymous record, use {|...|} instead of {...}.")
```

File: src/Compiler/FSComp.txt (hunk 4 above)
Comment (edgarfgp) [reply src/Compiler/FSComp.txt]: I mean they are related as it is trying to update an annon record using the record syntax. and we now that because 2 of the fields belong to an annon record. ?

File: tests/FSharp.Compiler.ComponentTests/Conformance/Types/RecordTypes/AnonymousRecords.fs (hunk 11 above)
Comment (edgarfgp) [reply tests/FSharp.Compiler.ComponentTests/Conformance/Types/RecordTypes/AnonymousRecords.fs:125]: I think so. Let me check if I have the range available

Comment (edgarfgp) [reply tests/FSharp.Compiler.ComponentTests/Conformance/Types/RecordTypes/AnonymousRecords.fs]: @0101 Ok took me some time but my last commit might do what was suggested. Please take a look and let me know if that is what you prefer and if the ranges are where it is required :)

Comment (0101) [reply tests/FSharp.Compiler.ComponentTests/Conformance/Types/RecordTypes/AnonymousRecords.fs]: Looks good! I guess we lost the diagnostic for invalid field (B), but I'm fine with that.

**Review by 0101 — APPROVED** [review]
